internal/exec/util/file.go

```diff
@@ -33,69 +36,123 @@ const (
 )
 
 type File struct {
-	Path     types.Path
-	Contents []byte
-	Mode     os.FileMode
-	Uid      int
-	Gid      int
+	io.ReadCloser
+	hash.Hash
+	Path         types.Path
+	Mode         os.FileMode
+	Uid          int
+	Gid          int
+	expectedHash string
 }
 
+func (f File) Verify() error {
+	if f.Hash == nil {
+		return nil
+	}
+	sum := f.Sum(nil)
+	encodedSum := make([]byte, hex.EncodedLen(len(sum)))
+	hex.Encode(encodedSum, sum)
+
+	if string(encodedSum) != f.expectedHash {
+		return util.ErrHashMismatch{
+			Calculated: string(encodedSum),
+			Expected:   f.expectedHash,
+		}
+	}
+	return nil
+}
+
+//returns a new ReadCloser that also writes to the provided hasher
+func newHashedReader(reader io.ReadCloser, hasher hash.Hash) io.ReadCloser {
+	return struct {
+		io.Reader
+		io.Closer
+	}{io.TeeReader(reader, hasher), reader}
+}
+
+// Returns a *File with a Reader that downloads, sha512's, and unzips the incoming data
+// Returns nil if f had invalid options. Errors reading/verifying/uncompressing the file will
+// present themselves when the Reader is actually read from
 func RenderFile(l *log.Logger, f types.File) *File {
-	var contents []byte
+	var reader io.ReadCloser
 	var err error
+	var expectedHash string
 
-	fetch := func() error {
-		contents, err = util.FetchResource(l, url.URL(f.Contents.Source))
-		return err
+	reader, err = util.FetchResourceAsReader(l, url.URL(f.Contents.Source))
+	if err != nil {
+		l.Crit("Error fetching file %q: %v", f.Path, err)
+		return nil
 	}
 
-	validate := func() error {
-		return util.AssertValid(f.Contents.Verification, contents)
+	hasher, err := util.GetHasher(f.Contents.Verification)
+	if err != nil {
+		l.Crit("Error verifying file %q: %v", f.Path, err)
+		return nil
 	}
 
-	decompress := func() error {
-		contents, err = decompressFile(l, f, contents)
-		return err
+	if hasher != nil {
+		reader = newHashedReader(reader, hasher)
+		expectedHash = f.Contents.Verification.Hash.Sum
 	}
 
-	if l.LogOp(fetch, "fetching file %q", f.Path) != nil {
+	reader, err = decompressFileStream(l, f, reader)
+	if err != nil {
+		l.Crit("Error decompressing file %q: %v", f.Path, err)
 		return nil
 	}
-	if l.LogOp(validate, "validating file contents") != nil {
-		return nil
+
+	return &File{
+		Path:         f.Path,
+		ReadCloser:   reader,
+		Hash:         hasher,
+		Mode:         os.FileMode(f.Mode),
+		Uid:          f.User.Id,
+		Gid:          f.Group.Id,
+		expectedHash: expectedHash,
 	}
-	if l.LogOp(decompress, "decompressing file contents") != nil {
-		return nil
+}
+
+// Wrapper for gzip's reader that closes the stream it wraps as well
+type gzipReader struct {
```

Comment:
The problem here is that gzip.NewReader() takes in a io.Reader but returns an io.ReadCloser (interestingly, it implements the io.ReadCloser interface, but the type signature of the function says it returns a io.Reader). Close() on the gzip.Reader does not actually close reader the it is wrapping, which is why I have the extra type that Close()'s both.